src/main/daml/Daml/Finance/Lifecycle/Rule/ClaimWithCompression.daml

```diff
@@ -0,0 +1,154 @@
+              consumedSteps = createSteps True consumedNetted holding
+              producedSteps = createSteps False producedNetted holding
+            pure $ consumedSteps <> producedSteps
+
+        -- Settlement steps
+        steps <- mconcat <$> sequence (zipWith calculateSteps holdings holdingCids)
+
+        -- Compress non-target-instrument legs
+        let
+          compressedSteps = uncurry (<>) $ compress <$>
+            partition (\s -> s.quantity.unit.id == effectView.targetInstrument.id) steps
+
+        -- Discover settlement routes
+        routedSteps <- exercise routeProviderCid RouteProvider.Discover with
+          discoverors = providers; contextId = None; steps = compressedSteps
+
+        -- Generate settlement instructions for other instruments
+        (batchCid, instructionCids) <- exercise settlementFactoryCid Settlement.Instruct with
+          instructors = providers
+          settlers
+          id = batchId
+          description = effectView.description
+          contextId = Some effectView.id
+          routedSteps
+          settlementTime = effectView.settlementTime
+
+        pure Claim.ClaimResult with batchCid; instructionCids
+
+-- Compress RoutedSteps by bypassing intermediaries.
+-- It only handled cycles such as A -> B, B -> A but does not handle
+-- cycles such as A -> B, B -> C, C -> A. The latter case can be handled
+-- by doing multiple passes of this function.
+compress : [Step] -> [Step]
+compress = compressRec []
+
+-- | HIDE
+-- Recursive utility for Step compression.
+-- The first argument is the list of steps for which we know that
+-- the receiver is not an intermediary.
+-- The second argument is the list of all other steps.
+-- At each stage, we
+-- - take the first element of the second list and consider the receiving party
+-- - look for steps where this party is an intermediary
+-- - if found, apply the bypass
+compressRec : [Step] -> [Step] -> [Step]
+compressRec visited [] = visited
+compressRec visited (target :: rest) =
+  let
+    (newtarget, newVisited) = foldr folder (target, []) visited
+    (newertarget, newRest)  = foldr folder (newtarget, []) rest
+    newerVisited = if (newertarget.quantity.amount == 0.0)
+      then newVisited
+      else newertarget :: newVisited
+  in
+    compressRec newerVisited newRest
+  where
+    folder elem (acc, rest) = (rest <>) <$> replace acc elem
+
+-- | HIDE
+-- Given two steps, check if they are adjacent.
+-- If they are adjacent, try to bypass the intermediary.
+-- Examples:
+-- A -- 10 --> B, C -- 10 --> D is left unchanged
+-- A -- 10 --> B, B -- 10 --> C is modified to A -- 0 --> B, A -- 10 --> C
+-- A -- 10 --> B, B -- 20 --> C is modified to A -- 0 --> B, B -- 10 --> C, A -- 10 --> C
+-- A -- 20 --> B, B -- 10 --> C is modified to A -- 10 --> B, A -- 10 --> C
+replace : Step -> Step -> (Step, [Step])
+replace s1 s2 | s1.receiver /= s2.sender
+              || s1.quantity.unit /= s2.quantity.unit
+              = (s1, [s2])
```

Comment:
Curious if there's an alternative to using a `guard` and then using an `or` predicate immediately below the guard... 🤔 

I get what you're doing here but I imagine this will catch some readers out.

```suggestion
replace s1 s2 | (s1.receiver /= s2.sender) || (s1.quantity.unit /= s2.quantity.unit) = (s1, [s2])
```

Reply:
My suggestion my be even more confusing 🤔

Reply:
Or : 
```suggestion
replace s1 s2 | (s1.receiver /= s2.sender || s1.quantity.unit /= s2.quantity.unit) = (s1, [s2])
```

Reply:
Thanks! I just inlined it